### PR TITLE
#43: client location

### LINKED ISSUE #43: Display client's IP address
* core task: #39  
* Display IP address and location.
* Also, provide refresh button. (Maybe refresh automatically with some interval instead of the button?)
* Update IP address & location along with clients table refresh (after any change). (GeneralUpdateEvent)
* Client's IP address should be retrieved with GET request to http://ip.cucurum.ru. The request should be executed on client side with JavaScript code. Example:
```java
getUI().ifPresent(ui -> ui.getPage().executeJs("""
                          var request = new XMLHttpRequest();
                          request.open("GET", "http://ip.cucurum.ru", false); // false for synchronous request
                          request.send(null);
                          return request.responseText;
                    """).then(String.class, System.out::println));
```
* Having received IP address, we can also get client's location using `LocationService.getLocation`
* Design 😄 :
IMG_20240907_181933

### PR DESCRIPTION
**Pull request by shurshurik**
Added text field with client's IP address and location

### CODE REVIEW

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.*;
+import java.util.concurrent.*;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class LocationUpdaterImpl implements LocationUpdater {
+    private final LocationService locationService;
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+    private final Map<UI, Set<Future<?>>> futures = new ConcurrentHashMap<>();
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]:
```suggestion
    private final Map<UI, Set<Future<?>>> scheduledTasksByUis = new ConcurrentHashMap<>();
```

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public void startScheduledUpdates(UI ui, LocationComponent locationComponent) {
+        var future = scheduler.scheduleAtFixedRate(
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]:
```suggestion
        var newScheduledTask = scheduler.scheduleAtFixedRate(
```

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                        log.debug("Error updating location", e);
+                        futures.get(ui).forEach(unusedFuture -> unusedFuture.cancel(false));
+                        futures.remove(ui);
+                    }
+                },
+                0,
+                5,
+                TimeUnit.SECONDS
+        );
+        futures.computeIfAbsent(ui, k -> ConcurrentHashMap.newKeySet()).add(future);
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]:
```suggestion
        futures.computeIfAbsent(ui, newUi -> ConcurrentHashMap.newKeySet()).add(future);
```

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    private final Map<UI, Set<Future<?>>> futures = new ConcurrentHashMap<>();
+
+    @Override
+    public void startScheduledUpdates(UI ui, LocationComponent locationComponent) {
+        var future = scheduler.scheduleAtFixedRate(
+                () -> {
+                    try {
+                        ui.access(() -> updateLocation(ui, locationComponent));
+                    } catch (Exception e) {
+                        log.debug("Error updating location", e);
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]:
```suggestion
                        log.debug("Exception while updating location", e);
```

**Review by KirillSmirnow — CHANGES_REQUESTED** [review]

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java (hunk 1 above)
Comment (shurshurik) [reply vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]: this and all subsequent proposals have been accepted)

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public void startScheduledUpdates(UI ui, LocationComponent locationComponent) {
+        var newScheduledTask = scheduler.scheduleAtFixedRate(
+                () -> {
+                    try {
+                        ui.access(() -> updateLocation(ui, locationComponent));
+                    } catch (Exception e) {
+                        log.debug("Exception while updating location", e);
+                        scheduledTasksByUis.get(ui).forEach(unusedFuture -> unusedFuture.cancel(false));
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]:
```suggestion
                        scheduledTasksByUis.get(ui).forEach(unusedTask -> unusedTask.cancel(false));
```

Comment (shurshurik) [reply vpn-router-web/src/main/java/vpnrouter/web/ui/clients/location/updater/LocationUpdaterImpl.java]: accepted)

**Review by KirillSmirnow — APPROVED** [review]

**Pull request merged**